src/Storages/MergeTree/MergeMutateExecutor.h

```diff
@@ -0,0 +1,160 @@
+#pragma once
+
+#include <deque>
+#include <functional>
+#include <atomic>
+#include <mutex>
+#include <condition_variable>
+
+#include <common/shared_ptr_helper.h>
+#include <Common/ThreadPool.h>
+#include <Storages/MergeTree/ExecutableTask.h>
+#include <Storages/MergeTree/MergeTreeData.h>
+
+
+namespace DB
+{
+
+class LambdaAdapter : public shared_ptr_helper<LambdaAdapter>, public ExecutableTask
+{
+public:
+
+    template <typename T>
+    explicit LambdaAdapter(T && inner_, MergeTreeData & data_) : inner(inner_), data(data_) {}
+
+    bool execute() override
+    {
+        res = inner();
+        inner = {};
+        return false;
+    }
+
+    void onCompleted() override
+    {
+        data.triggerBackgroundOperationTask(!res);
+    }
+
+    StorageID getStorageID() override
+    {
+        return data.getStorageID();
+    }
+
+private:
+    bool res = false;
+    std::function<bool()> inner;
+    MergeTreeData & data;
+};
+
+
+class MergeTreeBackgroundExecutor : public shared_ptr_helper<MergeTreeBackgroundExecutor>
+{
+public:
+
+    using CountGetter = std::function<size_t()>;
+    using Callback = std::function<void()>;
+
+
+    MergeTreeBackgroundExecutor()
+    {
+        scheduler = ThreadFromGlobalPool([this]() { schedulerThreadFunction(); });
+    }
+
+    ~MergeTreeBackgroundExecutor()
+    {
+        wait();
+    }
+
+    void setThreadsCount(CountGetter && getter)
+    {
+        threads_count_getter = getter;
+    }
+
+    void setTasksCount(CountGetter && getter)
+    {
+        max_task_count_getter = getter;
+    }
+
+    void setMetric(CurrentMetrics::Metric metric_)
+    {
+        metric = metric_;
+    }
+
+    bool trySchedule(ExecutableTaskPtr task)
+    {
+        std::lock_guard lock(mutex);
+
+        if (shutdown_suspend)
+            return false;
+
+        auto & value = CurrentMetrics::values[metric];
+        if (value.load() >= static_cast<int64_t>(max_task_count_getter()))
+            return false;
+
+        CurrentMetrics::add(metric);
+
+        tasks.emplace_back(task);
+        has_tasks.notify_one();
+        return true;
+    }
+
+    void removeTasksCorrespondingToStorage(StorageID id);
+
+    void wait()
+    {
+        {
+            std::lock_guard lock(mutex);
+            shutdown_suspend = true;
+            has_tasks.notify_all();
+        }
+
+        if (scheduler.joinable())
+            scheduler.join();
+
+        pool.wait();
+    }
+
+    size_t active()
+    {
+        return pool.active();
+    }
+
+    size_t pending()
+    {
+        std::lock_guard lock(mutex);
+        return tasks.size();
+    }
+
+private:
+
+    void updatePoolConfiguration()
+    {
+        const auto max_threads = threads_count_getter();
+        pool.setMaxFreeThreads(0);
```

Comment:
I really hope this stuff works.
Let's at least add a check that max_threads > 0

Reply:
This is needed to release all threads when there is nothing to do. This logic was in previous version of `BackgroundJobsExecutor`.